crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides an API for the accountable threshold multi-sig PureEdDSA signature scheme
+//! over the ed25519 twisted Edwards curve as defined in [RFC8032](https://tools.ietf.org/html/rfc8032).
+//!
+//! Signature verification also checks and rejects non-canonical signatures.
+
+use crate::ed25519::{
+    Ed25519PrivateKey, Ed25519PublicKey, Ed25519Signature, ED25519_PRIVATE_KEY_LENGTH,
+    ED25519_PUBLIC_KEY_LENGTH, ED25519_SIGNATURE_LENGTH,
+};
+use crate::{traits::*, HashValue};
+use anyhow::{anyhow, Result};
+use bit_vec::BitVec;
+use core::convert::TryFrom;
+use libra_crypto_derive::{DeserializeKey, SerializeKey, SilentDebug, SilentDisplay};
+use once_cell::sync::Lazy;
+use std::fmt;
+
+const MAX_NUM_OF_KEYS: usize = 32;
+const BITVEC_NUM_OF_BYTES: usize = MAX_NUM_OF_KEYS / 8;
+static BITVEC_FIRST_KEY_SET: Lazy<BitVec> =
+    Lazy::new(|| BitVec::from_bytes(&[0b1000_0000, 0b0000_0000, 0b0000_0000, 0b0000_0000]));
+
+/// Vector of private keys in the multi-key Ed25519 structure along with the threshold.
+#[derive(DeserializeKey, SilentDisplay, SilentDebug, SerializeKey)]
+pub struct MultiEd25519PrivateKey(Vec<Ed25519PrivateKey>, u8);
+
+#[cfg(feature = "assert-private-keys-not-cloneable")]
+static_assertions::assert_not_impl_any!(MultiEd25519PrivateKey: Clone);
+
+/// Vector of public keys in the multi-key Ed25519 structure along with the threshold.
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct MultiEd25519PublicKey(pub(crate) Vec<Ed25519PublicKey>, pub(crate) u8);
+
+/// Vector of the multi-key signatures along with their index required to map signatures with
+/// their corresponding public keys.
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct MultiEd25519Signature(pub(crate) Vec<Ed25519Signature>, pub(crate) BitVec);
+
+impl MultiEd25519PrivateKey {
```

Comment:
This is a little confusing to me.... I understand a multi public key... but not a private multikey, we should each have our own and separated.

Reply:
I tried to originally avoid it, but it's required due to traits. In short, we define private key material per public key. Unlikely that someone will use it, but did it for consistency. In theory the new scheme does not require private keys, that's true and specific to this particular scheme!
Any ideas @huitseeker?

Reply:
If this is required, it might be best to panic if this code ever gets called... alternatively could we just claim that the private material is the regular Ed25519PrivateKey?

Reply:
It's possible to decouple the Private and (Signature, Public) types, in fact I believe they were decoupled a long time ago. But I don't believe the work pays for itself. See below for a solution.

> alternatively could we just claim that the private material is the regular Ed25519PrivateKey?

No, the types are designed to prevent mismatch between their corresponding public / private / signature parts.
But you can certainly design a `MultiEd25519PrivateKey` newtype that wraps around a single `Ed25519PrivateKey`. In fact I think that's the better approach.